From 24e805b9c5984fb8c61f4d978f5a3132232c6f33 Mon Sep 17 00:00:00 2001
From: Masatake Iwasaki <iwasakims@apache.org>
Date: Sat, 13 Feb 2016 03:31:15 +0900
Subject: [PATCH 1372/1380] HDFS-9700. DFSClient and DFSOutputStream should
 set TCP_NODELAY on sockets for
 DataTransferProtocol (Gary Helmling via
 iwasakims)

(cherry picked from commit 372d1302c63c6f49f99be5766c5da9647ebd9ca6)

Conflicts:
	hadoop-hdfs-project/hadoop-hdfs-client/src/main/java/org/apache/hadoop/hdfs/client/impl/DfsClientConf.java

(cherry picked from commit f71fa500f9a43d928fbd27c873544898bff50dd4)
(cherry picked from commit 2dd80938a287416807af29238cb9a831085574e1)

Conflicts:
	hadoop-hdfs-project/hadoop-hdfs-client/src/main/java/org/apache/hadoop/hdfs/DFSClient.java
	hadoop-hdfs-project/hadoop-hdfs-client/src/main/java/org/apache/hadoop/hdfs/DataStreamer.java
	hadoop-hdfs-project/hadoop-hdfs-client/src/main/java/org/apache/hadoop/hdfs/client/HdfsClientConfigKeys.java
	hadoop-hdfs-project/hadoop-hdfs-client/src/main/java/org/apache/hadoop/hdfs/client/impl/DfsClientConf.java

Change-Id: I83e059991356b27cba616cd02516ca1f3d8dd8a4
(cherry picked from commit e479e9be1176079a4f327876058136d4be0f2eda)
---
 .../java/org/apache/hadoop/hdfs/DFSClient.java     |   15 +++++++++++++++
 .../java/org/apache/hadoop/hdfs/DFSConfigKeys.java |    5 +++++
 .../org/apache/hadoop/hdfs/DFSOutputStream.java    |    1 +
 3 files changed, 21 insertions(+)

diff --git a/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/DFSClient.java b/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/DFSClient.java
index e64223b..49102de 100644
--- a/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/DFSClient.java
+++ b/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/DFSClient.java
@@ -59,6 +59,8 @@
 import static org.apache.hadoop.hdfs.DFSConfigKeys.DFS_CLIENT_WRITE_PACKET_SIZE_DEFAULT;
 import static org.apache.hadoop.hdfs.DFSConfigKeys.DFS_CLIENT_WRITE_PACKET_SIZE_KEY;
 import static org.apache.hadoop.hdfs.DFSConfigKeys.DFS_DATANODE_SOCKET_WRITE_TIMEOUT_KEY;
+import static org.apache.hadoop.hdfs.DFSConfigKeys.DFS_DATA_TRANSFER_CLIENT_TCPNODELAY_DEFAULT;
+import static org.apache.hadoop.hdfs.DFSConfigKeys.DFS_DATA_TRANSFER_CLIENT_TCPNODELAY_KEY;
 import static org.apache.hadoop.hdfs.DFSConfigKeys.DFS_REPLICATION_DEFAULT;
 import static org.apache.hadoop.hdfs.DFSConfigKeys.DFS_REPLICATION_KEY;
 
@@ -340,6 +342,8 @@
     private final List<Class<? extends ReplicaAccessorBuilder>>
       replicaAccessorBuilderClasses;
 
+    private final boolean dataTransferTcpNoDelay;
+
     public Conf(Configuration conf) {
       // The hdfsTimeout is currently the same as the ipc timeout 
       hdfsTimeout = Client.getTimeout(conf);
@@ -365,6 +369,9 @@ public Conf(Configuration conf) {
           CommonConfigurationKeysPublic.IO_FILE_BUFFER_SIZE_KEY,
           CommonConfigurationKeysPublic.IO_FILE_BUFFER_SIZE_DEFAULT);
       defaultChecksumOpt = getChecksumOptFromConf(conf);
+      dataTransferTcpNoDelay = conf.getBoolean(
+          DFS_DATA_TRANSFER_CLIENT_TCPNODELAY_KEY,
+          DFS_DATA_TRANSFER_CLIENT_TCPNODELAY_DEFAULT);
       socketTimeout = conf.getInt(DFS_CLIENT_SOCKET_TIMEOUT_KEY,
           HdfsServerConstants.READ_TIMEOUT);
       /** dfs.write.packet.size is an internal config variable */
@@ -600,6 +607,13 @@ private DataChecksum createChecksum(ChecksumOpt userOpt) {
       }
       return dataChecksum;
     }
+
+    /**
+     * @return whether TCP_NODELAY should be set on client sockets
+     */
+    public boolean getDataTransferTcpNoDelay() {
+      return dataTransferTcpNoDelay;
+    }
   }
  
   public Conf getConf() {
@@ -2380,6 +2394,7 @@ private IOStreamPair connectToDN(DatanodeInfo dn, int timeout,
         LOG.debug("Connecting to datanode " + dnAddr);
       }
       NetUtils.connect(sock, NetUtils.createSocketAddr(dnAddr), timeout);
+      sock.setTcpNoDelay(getConf().getDataTransferTcpNoDelay());
       sock.setSoTimeout(timeout);
   
       OutputStream unbufOut = NetUtils.getOutputStream(sock);
diff --git a/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/DFSConfigKeys.java b/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/DFSConfigKeys.java
index 60b5914..e5faa6f 100644
--- a/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/DFSConfigKeys.java
+++ b/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/DFSConfigKeys.java
@@ -168,6 +168,11 @@
   public static final String  DFS_DATANODE_DUPLICATE_REPLICA_DELETION = "dfs.datanode.duplicate.replica.deletion";
   public static final boolean DFS_DATANODE_DUPLICATE_REPLICA_DELETION_DEFAULT = true;
 
+  public static final String DFS_DATA_TRANSFER_CLIENT_TCPNODELAY_KEY =
+      "dfs.data.transfer.client.tcpnodelay";
+  public static final boolean DFS_DATA_TRANSFER_CLIENT_TCPNODELAY_DEFAULT =
+      true;
+
   public static final String  DFS_NAMENODE_PATH_BASED_CACHE_BLOCK_MAP_ALLOCATION_PERCENT =
     "dfs.namenode.path.based.cache.block.map.allocation.percent";
   public static final float    DFS_NAMENODE_PATH_BASED_CACHE_BLOCK_MAP_ALLOCATION_PERCENT_DEFAULT = 0.25f;
diff --git a/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/DFSOutputStream.java b/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/DFSOutputStream.java
index 33e34c7..c4fb42b 100644
--- a/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/DFSOutputStream.java
+++ b/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/DFSOutputStream.java
@@ -1780,6 +1780,7 @@ static Socket createSocketForPipeline(final DatanodeInfo first,
     final Socket sock = client.socketFactory.createSocket();
     final int timeout = client.getDatanodeReadTimeout(length);
     NetUtils.connect(sock, isa, client.getRandomLocalInterfaceAddr(), client.getConf().socketTimeout);
+    sock.setTcpNoDelay(client.getConf().getDataTransferTcpNoDelay());
     sock.setSoTimeout(timeout);
     sock.setSendBufferSize(HdfsConstants.DEFAULT_DATA_SOCKET_SIZE);
     if(DFSClient.LOG.isDebugEnabled()) {
-- 
1.7.9.5

